File: jar/src/main/java/jline/api/aoi/Aoi_dist2ph.java
/**
 * @file Distribution conversion for AoI analysis
 *
 * @since LINE 3.0
 */
package jline.api.aoi;

import jline.util.matrix.Matrix;
import jline.util.matrix.MatrixCell;

public final class Aoi_dist2ph {
    private Aoi_dist2ph() {}

    /**
     * Convert LINE process representation to PH format for AoI analysis.
     */
    public static Aoi_dist2phResult aoi_dist2ph(MatrixCell proc) {
        if (proc.size() < 2) {
            throw new IllegalArgumentException("proc must be a cell array {D0, D1}");
        }

        Matrix D0 = proc.get(0);
        Matrix D1 = proc.get(1);

        if (D0.hasNaN() || D1.hasNaN()) {
            throw new IllegalArgumentException("Process contains NaN - class may be disabled");
        }

        int n = D0.getNumRows();
        if (n != D0.getNumCols() || n != D1.getNumRows() || n != D1.getNumCols()) {
            throw new IllegalArgumentException("D0 and D1 must be square matrices of the same size");
        }

        Matrix T = D0.copy();

        Matrix Q = D0.add(D1);

        for (int i = 0; i < n; i++) {
            double rowSum = 0.0;
            for (int j = 0; j < n; j++) {
                rowSum += Q.get(i, j);
            }
            if (Math.abs(rowSum) > 1e-10) {
                Q.set(i, i, Q.get(i, i) - rowSum);
            }
        }

        Matrix A = new Matrix(n + 1, n);
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                A.set(j, i, Q.get(i, j));
            }
        }
        for (int j = 0; j < n; j++) {
            A.set(n, j, 1.0);
        }

        Matrix b = new Matrix(n + 1, 1);
        b.set(n, 0, 1.0);

        Matrix At = A.transpose();
        Matrix AtA = At.mult(A);
        Matrix Atb = At.mult(b);
        Matrix piCol = new Matrix(n, 1);
        Matrix.solveSafe(AtA, Atb, piCol);

        Matrix pi = new Matrix(1, n);
        double piSum = 0.0;
        for (int j = 0; j < n; j++) {
            double v = Math.max(piCol.get(j, 0), 0.0);
            pi.set(0, j, v);
            piSum += v;
        }
        if (piSum > 0) {
            for (int j = 0; j < n; j++) {
                pi.set(0, j, pi.get(0, j) / piSum);
            }
        }

        Matrix ones = Matrix.ones(n, 1);
        Matrix completionRates = D1.mult(ones);

        Matrix alpha = new Matrix(1, n);
        double alphaSum = 0.0;
        for (int j = 0; j < n; j++) {
            double v = pi.get(0, j) * completionRates.get(j, 0);
            alpha.set(0, j, v);
            alphaSum += v;
        }

        if (alphaSum > 0) {
            for (int j = 0; j < n; j++) {
                alpha.set(0, j, alpha.get(0, j) / alphaSum);
            }
        } else {
            for (int j = 0; j < n; j++) {
                alpha.set(0, j, pi.get(0, j));
            }
        }

        double finalSum = 0.0;
        for (int j = 0; j < n; j++) {
            finalSum += alpha.get(0, j);
        }
        if (finalSum > 0 && Math.abs(finalSum - 1.0) > 1e-14) {
            for (int j = 0; j < n; j++) {
                alpha.set(0, j, alpha.get(0, j) / finalSum);
            }
        }

        for (int i = 0; i < n; i++) {
            if (T.get(i, i) > 0) {
                throw new IllegalArgumentException("T(" + i + "," + i + ") = " + T.get(i, i) + " > 0, expected non-positive");
            }
        }

        return new Aoi_dist2phResult(alpha, T);
    }
}

File: matlab/src/api/aoi/aoi_dist2ph.m
function [alpha, T] = aoi_dist2ph(proc)
%AOI_DIST2PH Convert LINE process representation to PH format for aoi-fluid
%
% [alpha, T] = AOI_DIST2PH(proc)
%
% Converts LINE's {D0, D1} MAP representation to (alpha, T) PH format
% where alpha is the initial probability vector and T is the sub-generator matrix.
%
% In PH representation:
%   - alpha: Row vector of initial probabilities (sums to 1)
%   - T: Sub-generator matrix (negative diagonal, non-negative off-diagonal)
%   - Absorption rates: -T * ones(n,1)
%
% Parameters:
%   proc (cell): LINE process representation {D0, D1} where:
%       D0: Sub-generator matrix (like T in PH)
%       D1: Completion/transition matrix
%
% Returns:
%   alpha: Initial probability row vector
%   T: Sub-generator matrix
%
% See also: aoi_extract_params, solveBufferless, solveSingleBuffer

% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

if isempty(proc) || ~iscell(proc) || length(proc) < 2
    error('aoi_dist2ph:InvalidInput', 'proc must be a cell array {D0, D1}');
end

D0 = proc{1};
D1 = proc{2};

% Check for NaN (disabled class)
if any(isnan(D0(:))) || any(isnan(D1(:)))
    error('aoi_dist2ph:DisabledClass', 'Process contains NaN - class may be disabled');
end

% Get dimensions
n = size(D0, 1);

if n ~= size(D0, 2) || n ~= size(D1, 1) || n ~= size(D1, 2)
    error('aoi_dist2ph:DimensionMismatch', 'D0 and D1 must be square matrices of the same size');
end

% T is the sub-generator (D0 in MAP notation)
T = D0;

% Compute stationary distribution of the underlying CTMC
% The full generator is Q = D0 + D1
Q = D0 + D1;

% Check if Q is a proper generator (rows should sum to ~0)
rowSums = sum(Q, 2);
if max(abs(rowSums)) > 1e-10
    % Q is not a proper generator, normalize
    for i = 1:n
        Q(i, i) = Q(i, i) - rowSums(i);
    end
end

% Compute stationary distribution pi of Q
% pi * Q = 0 and sum(pi) = 1
% Solve [Q'; ones(1,n)] * [pi'; 1] = [zeros(n,1); 1]
A = [Q'; ones(1, n)];
b = [zeros(n, 1); 1];
pi = (A \ b)';

% Ensure pi is non-negative and normalized
pi = max(pi, 0);
pi = pi / sum(pi);

% Initial probability alpha is the stationary distribution weighted by
% the probability of starting in each phase after a completion
% For a renewal process: alpha = pi * D1 / (pi * D1 * ones)
% This gives the probability of starting in each phase
completionRates = D1 * ones(n, 1);
numerator = pi .* completionRates';

if sum(numerator) > 0
    alpha = numerator / sum(numerator);
else
    % Fallback: use stationary distribution
    alpha = pi;
end

% Ensure alpha is a row vector
alpha = alpha(:)';

% Normalize to ensure sum = 1
alpha = alpha / sum(alpha);

% Validate T is a proper sub-generator
% - Diagonal elements should be non-positive
% - Off-diagonal elements should be non-negative
% - Row sums should be non-positive (not zero, since it's a sub-generator)
for i = 1:n
    if T(i, i) > 0
        error('aoi_dist2ph:InvalidSubGenerator', 'T(%d,%d) = %g > 0, expected non-positive', i, i, T(i, i));
    end
end

end
